pkg/reconciler/knative/resources/knative_configuration.go

```diff
@@ -18,17 +18,27 @@ package resources
 
 import (
 	"fmt"
-
 	"github.com/knative/serving/pkg/apis/autoscaling"
 	knservingv1alpha1 "github.com/knative/serving/pkg/apis/serving/v1alpha1"
 	"github.com/knative/serving/pkg/apis/serving/v1beta1"
 	"github.com/kubeflow/kfserving/pkg/apis/serving/v1alpha1"
 	"github.com/kubeflow/kfserving/pkg/constants"
 	"github.com/kubeflow/kfserving/pkg/utils"
+	"k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-func CreateKnativeConfiguration(name string, metadata metav1.ObjectMeta, modelSpec *v1alpha1.ModelSpec) *knservingv1alpha1.Configuration {
+type ConfigurationBuilder struct {
+	config *v1.ConfigMap
+}
+
+func NewConfigurationBuilder(config *v1.ConfigMap) *ConfigurationBuilder {
```

Comment:
I think if you pass the configuration to the spec in the ApplyDefaults call - you wont need any of these changes!

Reply:
Actually `ApplyDefault` is only supposed to mutate the fields on kfservice spec, but here we are setting values from configmap for container image.

Reply:
I am trying to figure out how to think about configmap mutations. If a configmap changes seems like there are 2 options: 
1. apply retrospective: mutate all existing services to conform to the new map
2. dont apply retrospective: new services get new config, old unchanged

Obviously (1) is more consistent, but scary to apply while (2) is easy to apply - but now 2 kfservices that look very similar can have very different implementations just based on when they were applied - IMO this is a terrible side effect.

This can be avoided if we have ApplyDefaults apply annotations or modify the spec (e.g. containerImage may be an annotation) and have the handlers use the annotations. This way I can inspect the kfservice to clearly see what is going on... (the only downside - that is arguably not a downside? is that we have more knobs... but annotations allow us to keep the spec simple?)

This also makes the implementation simple again :)

Reply:
I think we are going for option 1 which is simple and easy to reason about, it may sound a bit scary but with knative safe rollout the traffic just does not shift if something goes wrong with new configuration change. With option 2 in case user tries to fix the configuration issue, since we are not applying to services in retrospectively way, the services with bad config will not get a chance to reconcile and may potentially leave in a bad state. Of course user has to be super careful when applying changes for these kind of global configs.

Reply:
Okay so config map change will trigger mass redeployment. Agree that it is consistent. Still scared, but I guess fear is good when it comes to global configs :)

Reply:
Keep in mind, this is *not* how tools like istio work. Any sidecar configs are setup in the pod init container and pods must be redeployed. I really don't like this, as it violates infrastructure as code. 

I'm happy to move forward with the mass redeploy approach, but I'm prepared to see customers complain about this ;).